tooling/nargo/src/ops/debug.rs

```diff
@@ -0,0 +1,216 @@
+        compilation_result,
+        &workspace_file_manager,
+        compile_options.deny_warnings,
+        compile_options.silence_warnings,
+    )
+    .map(|compiled_program| transform_program(compiled_program, expression_width))
+}
+
+pub fn compile_options_for_debugging(
+    acir_mode: bool,
+    skip_instrumentation: bool,
+    expression_width: Option<ExpressionWidth>,
+    compile_options: CompileOptions,
+) -> CompileOptions {
+    CompileOptions {
+        // Compilation warnings are disabled when
+        // compiling for debugging
+        //
+        // For instrumenting the program the debugger
+        // will import functions that may not be used,
+        // which would generate compilation warnings
+        silence_warnings: true,
+        deny_warnings: false,
+        instrument_debug: !skip_instrumentation,
+        force_brillig: !acir_mode,
+        expression_width,
+        ..compile_options
+    }
+}
+
+pub fn prepare_package_for_debug<'a>(
+    file_manager: &'a FileManager,
+    parsed_files: &'a mut ParsedFiles,
+    package: &'a Package,
+    workspace: &Workspace,
+) -> (Context<'a, 'a>, CrateId) {
+    let debug_instrumenter = instrument_package_files(parsed_files, file_manager, package);
+
+    // -- This :down: is from nargo::ops(compile).compile_program_with_debug_instrumenter
+    let (mut context, crate_id) = prepare_package(file_manager, parsed_files, package);
+    link_to_debug_crate(&mut context, crate_id);
+    context.debug_instrumenter = debug_instrumenter;
+    context.package_build_path = workspace.package_build_path(package);
+    (context, crate_id)
+}
+
+pub fn load_workspace_files(workspace: &Workspace) -> (FileManager, ParsedFiles) {
+    let mut file_manager = file_manager_with_stdlib(Path::new(""));
+    insert_all_files_for_workspace_into_file_manager(workspace, &mut file_manager);
+
+    let parsed_files = parse_all(&file_manager);
+    (file_manager, parsed_files)
+}
+
+/// Add debugging instrumentation to all parsed files belonging to the package
+/// being compiled
+fn instrument_package_files(
+    parsed_files: &mut ParsedFiles,
+    file_manager: &FileManager,
+    package: &Package,
+) -> DebugInstrumenter {
+    // Start off at the entry path and read all files in the parent directory.
+    let entry_path_parent = package
+        .entry_path
+        .parent()
+        .unwrap_or_else(|| panic!("The entry path is expected to be a single file within a directory and so should have a parent {:?}", package.entry_path));
+
+    let mut debug_instrumenter = DebugInstrumenter::default();
+
+    for (file_id, parsed_file) in parsed_files.iter_mut() {
+        let file_path =
+            file_manager.path(*file_id).expect("Parsed file ID not found in file manager");
+        for ancestor in file_path.ancestors() {
+            if ancestor == entry_path_parent {
+                // file is in package
+                debug_instrumenter.instrument_module(&mut parsed_file.0, *file_id);
```

Comment:
Maybe add a `break` after this to avoid checking the ancestor parents, which won't match?
Alternatively, we could instrument if any if the ancestors match using `file_path.ancestors().iter().any(...)` which might be clearer here.